src/blob.c

```diff
@@ -137,9 +139,131 @@ Blob_diff_to_buffer(Blob *self, PyObject *args, PyObject *kwds)
     return wrap_patch(patch, self, NULL);
 }
 
+
+struct blob_filter_stream {
+    git_writestream stream;
+    PyObject *py_fdst;
+};
+
+static int blob_filter_stream_write(
+    git_writestream *s, const char *buffer, size_t len)
+{
+    struct blob_filter_stream *stream = (struct blob_filter_stream *)s;
+    PyObject *py_size;
+    PyGILState_STATE gil = PyGILState_Ensure();
+    int err = 0;
+
+    py_size = PyObject_CallMethod(stream->py_fdst, "write", "y#", buffer, len);
+    if (!py_size)
+    {
+        git_error_set(GIT_ERROR_OS, "could not write to pygit2 stream");
+        err = GIT_ERROR;
+    }
+    Py_DECREF(py_size);
+    PyGILState_Release(gil);
+    return err;
+}
+
+static int blob_filter_stream_close(git_writestream *s)
+{
+    return 0;
+}
+
+static void blob_filter_stream_free(git_writestream *s)
+{
+    struct blob_filter_stream *stream = (struct blob_filter_stream *)s;
+    Py_DECREF(stream->py_fdst);
+    stream->py_fdst = NULL;
+}
+
+
+PyDoc_STRVAR(Blob_filter__doc__,
+  "filter(fdst: BinaryIO, as_path: str, flags: int = GIT_BLOB_FILTER_CHECK_FOR_BINARY[, commit_id: oid = None]) -> None\n"
+  "\n"
+  "Copy the filtered contents of the blob to the file-like object `fdst`.\n"
+  "This applies filters as if the blob was being checked out to the\n"
+  "working directory under the specified filename.\n"
+  "\n"
+  "Returns: The filtered content.\n"
+  "\n"
+  "Parameters:\n"
+  "\n"
+  "as_path : str\n"
+  "    Treat blob as if it had this filename (used for attribute lookups).\n"
+  "\n"
+  "flags\n"
+  "    GIT_BLOB_FILTER_* bitflags.\n"
+  "\n"
+  "commit_id: oid\n"
+  "    Commit to load attributes from when\n"
+  "    GIT_BLOB_FILTER_ATTRIBUTES_FROM_COMMIT is specified.\n");
```

Comment:
Updated this to write to a file-like object instead of returning a bytestring.

With filtered content it's more likely that the resulting data will be large (especially in the LFS use case) and we may not want to or be able to load the entire object into memory (which is what `git_blob_filter()` does).

Using a stream object here allows us to do buffered writes (and the user can still get the full bytestring in memory if they want via `io.BytesIO`)

Reply:
I think I will probably end up reworking this again so that it returns something like an `io.BufferedReader` instance where the caller can `read()` when they want to instead of needing to write the full filtered blob into an intermediate file first